### PR TITLE
Add cross selling feature

### PR DESCRIPTION
**Pull request by patrickweh**

### CODE REVIEW

File: src/Actions/Product/CreateProduct.php [hunk 1]
```diff
@@ -54,6 +56,13 @@ public function performAction(): Product
             $product->prices()->createMany($prices);
         }
 
+        if ($productCrossSellings) {
```
Comment (SirSplasch) [inline src/Actions/Product/CreateProduct.php:59]: if statement unnecessary, foreach is enough because its defaulted to empty array. Its only an issue if you try to loop through something that isn't countable.

File: src/Actions/Product/CreateProduct.php [hunk 2]
```diff
@@ -54,6 +56,13 @@ public function performAction(): Product
...
+            foreach ($productCrossSellings as $productCrossSelling) {
+                $productCrossSelling['product_id'] = $product->id;
+                CreateProductCrossSelling::make($productCrossSelling)->checkPermission()->validate()->execute()->id;
```
Comment (SirSplasch) [inline src/Actions/Product/CreateProduct.php:62]: if you want to check the permission do this beforehand, outside the foreach loop. In this case for each productCrossSelling the same permission is checked although its extremly unlikely that the users permissions do change while looping. And even if they change, this can cause an unexpected behaviour for the user.

Comment (SirSplasch) [inline src/Actions/Product/CreateProduct.php:62]: You know that this will eventually throw an exception. This may cause some confusion, because the product is created, but an exception is thrown. Rethink if you need to validate. Also the permission should be checked before anything happens with the product to avoid confusion.

Comment (SirSplasch) [inline src/Actions/Product/CreateProduct.php:62]: remove ->id doesn't do anything

File: src/Actions/Product/UpdateProduct.php [hunk 3]
```diff
@@ -85,6 +90,30 @@ public function performAction(): Model
...
+            $activeProductCrossSellings = [];
+            foreach ($productCrossSellings as $productCrossSelling) {
+                $productCrossSelling['product_id'] = $product->id;
+                if ($productCrossSelling['id'] ?? false) {
+                    $action = UpdateProductCrossSelling::make($productCrossSelling);
+                } else {
+                    $action = CreateProductCrossSelling::make($productCrossSelling);
+                }
+
+                $activeProductCrossSellings[] = $action->checkPermission()->validate()->execute()->id;
```
Comment (SirSplasch) [inline src/Actions/Product/UpdateProduct.php:103]: same as always, rethink if its that smart if an exception is thrown at this point in the code.

File: src/Actions/Product/UpdateProduct.php [hunk 4]
```diff
@@ -85,6 +90,30 @@ public function performAction(): Model
...
+            }
+
+            $removedProductCrossSellings = $product->productCrossSellings
+                ?->whereNotIn('id', $activeProductCrossSellings)
+                ->pluck('id')
+                ->toArray();
+            foreach ($removedProductCrossSellings as $removedProductCrossSelling) {
+                DeleteProductCrossSelling::make(['id' => $removedProductCrossSelling])
+                    ->checkPermission()
```
Comment (SirSplasch) [inline src/Actions/Product/UpdateProduct.php:112]: really need to check the permission for every deletion? Also if an exception is thrown this can cause some confusion for the user, because the product is already updated at this point.

File: src/Helpers/PriceHelper.php
Comment (SirSplasch) [inline src/Helpers/PriceHelper.php:1]: why these changes? For my understanding the original code was perfectly fine.
Is it because you want isCalculated set to true only if there is no direct price from product in price list?

Even if this is the case, you don't need to copy the price into an extra variable, just set a variable with value true.

File: src/Http/Livewire/Product/Product.php [hunk 5]
```diff
@@ -120,6 +124,20 @@ public function save(): bool
                 ->toArray();
         }
 
+        if ($this->productCrossSellings !== null) {
+            $this->product['product_cross_sellings'] = collect($this->productCrossSellings)
```
Comment (SirSplasch) [inline src/Http/Livewire/Product/Product.php:128]: just use array_map() no need to collect()->toArray()

File: src/Http/Livewire/Product/Product.php [hunk 6]
```diff
@@ -120,6 +124,20 @@ public function save(): bool
...
+                ->map(function (array $productCrossSelling) {
+                    $productCrossSelling['products'] = collect($productCrossSelling['products'])
```
Comment (SirSplasch) [inline src/Http/Livewire/Product/Product.php:130]: same here again, just user array_map()

File: src/Http/Livewire/Product/Product.php [hunk 7]
```diff
@@ -120,6 +124,20 @@ public function save(): bool
...
+                        ->map(function (array $product) {
```
Comment (SirSplasch) [inline src/Http/Livewire/Product/Product.php:131]: callback can be one-lined fn ($product) => $product['id']

**Review by SirSplasch — CHANGES_REQUESTED** [review]

File: src/Helpers/PriceHelper.php
Comment (patrickweh) [reply src/Helpers/PriceHelper.php:1]: the original code added isCalculated = true even if it was a "real" price model with an database entry.
i need to know if the price is real or just virutal, the change works fine for this purpose

File: src/Http/Requests/UpdateProductCrossSellingRequest.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace FluxErp\Http\Requests;
+
+use FluxErp\Rules\ExistsWithIgnore;
+use Illuminate\Foundation\Http\FormRequest;
```
Comment (SirSplasch) [inline src/Http/Requests/UpdateProductCrossSellingRequest.php:6]: remove unused import

File: src/Actions/Product/UpdateProduct.php [hunk 9]
```diff
@@ -85,6 +91,35 @@ public function performAction(): Model
...
+                $productCrossSelling['product_id'] = $product->id;
+                if ($productCrossSelling['id'] ?? false) {
+                    $action = UpdateProductCrossSelling::make($productCrossSelling);
+                } else {
+                    $action = CreateProductCrossSelling::make($productCrossSelling);
+                }
+
+                try {
+                    $activeProductCrossSellings[] = $action->checkPermission()->validate()->execute();
+                } catch (ValidationException) {}
```
Comment (SirSplasch) [inline src/Actions/Product/UpdateProduct.php:106]: also catch UnauthorizedException

**Review by SirSplasch — CHANGES_REQUESTED** [review]

**Review by SirSplasch — APPROVED** [review]
